src/Equation.jl

```diff
@@ -178,6 +192,48 @@ function copy_node(tree::Node{T})::Node{T} where {T}
     end
 end
 
+"""
+    copy_node_with_topology(
+        tree::Node{T}, id_map::IdDict{Node{T},Node{T}}
+    )::Node{T} where {T}
+
+id_map is a map from `objectid(tree)` to `copy(tree)`.
+We check against the map before making a new copy; otherwise
+we can simply reference the existing copy.
+[Thanks to Ted Hopp](https://stackoverflow.com/questions/49285475/how-to-copy-a-full-non-binary-tree-including-loops)
+"""
+function copy_node_with_topology(
+    tree::Node{T}, id_map::IdDict{Node{T},Node{T}}
+)::Node{T} where {T}
+
+    # This preserves nodes which have multiple parents.
+    # Otherwise, we would have the branch duplicated.
+    if haskey(id_map, tree)
+        return id_map[tree]
+    end
+
+    if tree.degree == 0
+        if tree.constant
+            copied_tree = Node(; val=copy(tree.val))
+        else
+            copied_tree = Node(T; feature=copy(tree.feature))
+        end
+        id_map[tree] = copied_tree
```

Comment:
Not sure if we need to enable multiple parents for a single leaf node - this part of the copies would be a bit more expensive. It is more generic though.

Reply:
I think I'll leave this in. What if a re-used expression is mutated from a tree into a single leaf node? We wouldn't want to break things...